File: landrop-main/sender/desktop/src/protocol/session.test.ts
import { describe, expect, it } from "vitest";

import {
  createSession,
  isSessionActive,
  isSessionExpired,
  sessionStatus,
} from "./session";
import type { DeviceInfo, Lifetime, SessionData } from "./types";

const fixtures: Record<string, unknown> = require(
  "../../../../protocol/fixtures/model-cases.json",
);

interface SessionCase {
  session_id: string;
  protocol_version: number;
  created_at_ms: number;
  expires_at_ms: number;
  now_ms: number;
  status: "not_started" | "active" | "expired";
}

const sessionCases: SessionCase[] = (fixtures.session_cases as unknown[]).map(
  (c) => c as SessionCase,
);

const asSession = (c: SessionCase): SessionData => ({
  session_id: c.session_id,
  protocol_version: c.protocol_version,
  sender: { display_name: "ISHAQ CYBERTECH" },
  receiver: { device_name: "Test Device" },
  state: "SESSION_ESTABLISHED",
  lifetime: { created_at_ms: c.created_at_ms, expires_at_ms: c.expires_at_ms },
});

const caseOf = (index: number, nowMs: number): SessionCase => {
  const c = sessionCases[index];
  if (c === undefined) throw new Error("missing session case");
  return { ...c, now_ms: nowMs };
};

describe("fixture parity - session status", () => {
  it.each(sessionCases)(
    "classifies now_ms=$now_ms against [$created_at_ms, $expires_at_ms) as $status",
    (c) => {
      expect(sessionStatus(asSession(c), c.now_ms)).toBe(c.status);
      expect(isSessionActive(asSession(c), c.now_ms)).toBe(c.status === "active");
      expect(isSessionExpired(asSession(c), c.now_ms)).toBe(c.status === "expired");
    },
  );
});

describe("sessionStatus boundaries", () => {
  it("is active through expires_at_ms - 1 and expired from expires_at_ms", () => {
    expect(sessionStatus(asSession(caseOf(0, 0)), 0)).toBe("active");
    expect(sessionStatus(asSession(caseOf(0, 999)), 999)).toBe("active");
    expect(sessionStatus(asSession(caseOf(0, 1000)), 1000)).toBe("expired");
  });

  it("is not_started strictly before created_at_ms", () => {
    expect(sessionStatus(asSession(caseOf(4, 100)), 100)).toBe("not_started");
    expect(sessionStatus(asSession(caseOf(4, 499)), 499)).toBe("not_started");
    expect(sessionStatus(asSession(caseOf(4, 500)), 500)).toBe("active");
  });

  it("is deterministic for times outside the window", () => {
    expect(sessionStatus(asSession(caseOf(0, -10)), -10)).toBe("not_started");
    expect(sessionStatus(asSession(caseOf(0, 1e15)), 1e15)).toBe("expired");
  });
});

describe("createSession", () => {
  const base = {
    session_id: "sess_01JTEST",
    protocol_version: 1,
    sender: { display_name: "ISHAQ CYBERTECH" },
    receiver: { device_name: "Test Device" } satisfies DeviceInfo,
    lifetime: { created_at_ms: 0, expires_at_ms: 1000 } satisfies Lifetime,
  };

  it("creates a session in the initial delivery state", () => {
    const result = createSession(base);
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(result.session).toEqual({
        session_id: "sess_01JTEST",
        protocol_version: 1,
        sender: base.sender,
        receiver: base.receiver,
        state: "DISCOVERING",
        lifetime: base.lifetime,
      });
    }
  });

  it("rejects an invalid session id", () => {
    const result = createSession({ ...base, session_id: "bad id!" });
    expect(result).toEqual({ ok: false, problem: "INVALID_SESSION_ID" });
  });

  it("rejects an unsupported protocol version", () => {
    const result = createSession({ ...base, protocol_version: 2 });
    expect(result).toEqual({ ok: false, problem: "UNSUPPORTED_PROTOCOL" });
  });

  it("rejects a zero-duration lifetime", () => {
    const result = createSession({
      ...base,
      lifetime: { created_at_ms: 100, expires_at_ms: 100 },
    });
    expect(result).toEqual({ ok: false, problem: "INVALID_LIFETIME" });
  });

  it("rejects an inverted lifetime", () => {
    const result = createSession({
      ...base,
      lifetime: { created_at_ms: 1000, expires_at_ms: 100 },
    });
    expect(result).toEqual({ ok: false, problem: "INVALID_LIFETIME" });
  });

  it("rejects non-integer times", () => {
    const fractional = createSession({
      ...base,
      lifetime: { created_at_ms: 0.5, expires_at_ms: 1000 },
    });
    expect(fractional).toEqual({ ok: false, problem: "INVALID_LIFETIME" });
    const nonInteger = createSession({
      ...base,
      lifetime: { created_at_ms: 0, expires_at_ms: 1000.1 },
    });
    expect(nonInteger).toEqual({ ok: false, problem: "INVALID_LIFETIME" });
  });

  it("rejects a negative created time", () => {
    const result = createSession({
      ...base,
      lifetime: { created_at_ms: -1, expires_at_ms: 1000 },
    });
    expect(result).toEqual({ ok: false, problem: "INVALID_LIFETIME" });
  });

  it("is platform-deterministic", () => {
    const a = createSession(base);
    const b = createSession(base);
    expect(JSON.stringify(a)).toBe(JSON.stringify(b));
  });
});

File: landrop-main/receiver/android/app/src/test/java/com/landrop/receiver/protocol/SessionTest.kt
package com.landrop.receiver.protocol

import org.junit.Assert.assertEquals
import org.junit.Test

class SessionTest {

    data class Case(
        val created: Long,
        val expires: Long,
        val now: Long,
        val status: SessionStatus,
    )

    private val cases = listOf(
        Case(0, 1000, 500, SessionStatus.ACTIVE),
        Case(0, 1000, 999, SessionStatus.ACTIVE),
        Case(0, 1000, 1000, SessionStatus.EXPIRED),
        Case(0, 1000, 2000, SessionStatus.EXPIRED),
        Case(500, 1000, 100, SessionStatus.NOT_STARTED),
        Case(500, 1000, 500, SessionStatus.ACTIVE),
    )

    private fun session(created: Long, expires: Long): SessionData =
        SessionData(
            SessionId("sess_01JTEST"),
            ProtocolVersion(1),
            SenderProfile("ISHAQ CYBERTECH"),
            DeviceInfo("Test Device"),
            DeliveryState.SESSION_ESTABLISHED,
            Lifetime(created, expires),
        )

    @Test
    fun fixtureCasesClassifyByTimeWindow() {
        for (c in cases) {
            val session = session(c.created, c.expires)
            assertEquals("now=${c.now} in [${c.created},${c.expires})", c.status, sessionStatus(session, c.now))
            assertEquals(c.status == SessionStatus.ACTIVE, isSessionActive(session, c.now))
            assertEquals(c.status == SessionStatus.EXPIRED, isSessionExpired(session, c.now))
        }
    }

    @Test
    fun boundarySemanticsAreExclusiveAtExpiry() {
        val session = session(0, 1000)
        assertEquals(SessionStatus.ACTIVE, sessionStatus(session, 0))
        assertEquals(SessionStatus.ACTIVE, sessionStatus(session, 999))
        assertEquals(SessionStatus.EXPIRED, sessionStatus(session, 1000))
    }

    @Test
    fun deterministicOutsideWindow() {
        val session = session(0, 1000)
        assertEquals(SessionStatus.NOT_STARTED, sessionStatus(session, -10))
        assertEquals(SessionStatus.EXPIRED, sessionStatus(session, 1_000_000_000_000_000))
    }

    private fun draft() = SessionDraft(
        sessionId = "sess_01JTEST",
        protocolVersion = 1,
        sender = SenderProfile("ISHAQ CYBERTECH"),
        receiver = DeviceInfo("Test Device"),
        lifetime = Lifetime(0, 1000),
    )

    @Test
    fun createsSessionInInitialState() {
        val result = createSession(draft()) as SessionCreateResult.Ok
        assertEquals(DeliveryState.DISCOVERING, result.session.state)
        assertEquals(SessionId("sess_01JTEST"), result.session.sessionId)
        assertEquals(ProtocolVersion(1), result.session.protocolVersion)
    }

    @Test
    fun rejectsInvalidSessionId() {
        assertEquals(
            SessionCreateResult.Failing(SessionProblem.INVALID_SESSION_ID),
            createSession(draft().copy(sessionId = "bad id!")),
        )
    }

    @Test
    fun rejectsUnsupportedProtocol() {
        assertEquals(
            SessionCreateResult.Failing(SessionProblem.UNSUPPORTED_PROTOCOL),
            createSession(draft().copy(protocolVersion = 2)),
        )
    }

    @Test
    fun rejectsZeroDurationLifetime() {
        assertEquals(
            SessionCreateResult.Failing(SessionProblem.INVALID_LIFETIME),
            createSession(draft().copy(lifetime = Lifetime(100, 100))),
        )
    }

    @Test
    fun rejectsInvertedLifetime() {
        assertEquals(
            SessionCreateResult.Failing(SessionProblem.INVALID_LIFETIME),
            createSession(draft().copy(lifetime = Lifetime(1000, 100))),
        )
    }

    @Test
    fun rejectsNegativeCreatedTime() {
        assertEquals(
            SessionCreateResult.Failing(SessionProblem.INVALID_LIFETIME),
            createSession(draft().copy(lifetime = Lifetime(-1, 1000))),
        )
    }
}